go/src/pluginmgr/pluginmgr_common/pluginmgr_helper.go

```diff
@@ -0,0 +1,505 @@
+	}
+
+	pluginMgr = &PluginManager{
+		clientTx:       clientTx,
+		plugins:        make(map[string]plugins_common.Plugin),
+		pluginMetadata: make(map[string]plugins_common.IPluginMetadata),
+		stopch:         make(chan struct{}),
+	}
+
+	if err := pluginMgr.clientTx.RegisterClient(ProcID); err != nil {
+		return nil, err
+	}
+
+	return pluginMgr, nil
+}
+
+// getPluginManager : Get plugin manager object
+func (plmgr *PluginManager) getPlugin(plgname string) (plugins_common.Plugin, bool) {
+	plugin, ok := plmgr.plugins[plgname]
+	return plugin, ok
+}
+
+// getPluginMetadata : Get plugin metadata object
+func (plmgr *PluginManager) getPluginMetadata(plgname string) (plugins_common.IPluginMetadata, bool) {
+	pluginmetadata, ok := plmgr.pluginMetadata[plgname]
+	return pluginmetadata, ok
+}
+
+/*
+ * Listens for server requests via golib. This call blocks.
+ *
+ * Input:
+ *  stop - stop channel used to stop the blocking receive server call
+ *
+ * Output:
+ *  none -
+ *
+ * Return:
+ *  error - error message or nil on success
+ */
+func (plmgr *PluginManager) run() error {
+	for {
+		select {
+		case <-plmgr.stopch:
+			lomcommon.LogInfo("Stopping plugin manager run loop")
+			return nil
+		default:
+			serverReq, err := plmgr.clientTx.RecvServerRequest()
+			if err != nil {
+				lomcommon.LogError("Error RecvServerRequest() : %s", err)
+				return err
+			}
+			if serverReq == nil {
+				return lomcommon.LogError("Error RecvServerRequest() : nil")
+			}
+
+			switch serverReq.ReqType {
+			case lomipc.TypeServerRequestAction:
+				actionReq, ok := serverReq.ReqData.(*lomipc.ActionRequestData)
+				if !ok {
+					lomcommon.LogError("Error in parsing ActionRequestData")
+					continue
+				}
+
+				plugin, ok := plmgr.getPlugin(actionReq.Action)
+				if !ok {
+					lomcommon.LogError("Plugin %s not found", actionReq.Action)
+					continue
+				}
+				lomcommon.LogInfo("Received action request for plugin %s", plugin.GetPluginID())
+
+				/* TODO: Goutham : Handle Request, Do error checks, pass HeartBEat channel, handle timeouts, handle heartbeats etc
+				goroutinetracker.Start("plg_mgr_Run_Action_"+actionReq.Action+"_"+lomcommon.GetUUID(),
+					func() {
+						hbchan := make(chan plugins_common.PluginHeartBeat)// TODO: Goutham ; Do we need buffered instead ??
+						res := plugin.Request(hbchan, actionReq)
+						plmgr.clientTx.SendServerResponse(&lomipc.MsgSendServerResponse{
+							ReqType: lomipc.TypeServerRequestAction,
+							ResData: res,
+						})
+					})
+				*/
+			case lomipc.TypeServerRequestShutdown:
+				// TODO: Goutham : handle shutdown, haqndle synchronously, do not listen on responses from plugins, send deregister with server
+			default:
+				lomcommon.LogError("Unknown server request type")
+			}
+		}
+	}
+	return nil
+}
+
+// TODO: Goutham : For all plugins call shutdoen, change plugin state, etc
+func (plmgr *PluginManager) shutdown() error {
+	if err := plmgr.clientTx.DeregisterClient(); err != nil {
```

Comment:
Don't block on failure. Just log it.

for each plugin
       shutdown  -- wait for a timeout
Exit upon timeout
Log error on any running go routines per tracker

Reply:
added note for future implementation